server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -0,0 +1,372 @@
+                }
+                long firstPrimaryTerm = getPrimaryTerm(firstTokens);
+                long secondPrimaryTerm = getPrimaryTerm(secondTokens);
+                if (firstPrimaryTerm != secondPrimaryTerm) {
+                    return firstPrimaryTerm > secondPrimaryTerm ? 1 : -1;
+                } else {
+                    long firstGeneration = getGeneration(firstTokens);
+                    long secondGeneration = getGeneration(secondTokens);
+                    if (firstGeneration != secondGeneration) {
+                        return firstGeneration > secondGeneration ? 1 : -1;
+                    } else {
+                        return getUuid(firstTokens).compareTo(getUuid(secondTokens));
+                    }
+                }
+            }
+        }
+
+        // Visible for testing
+        static String getMetadataFilename(long primaryTerm, long generation, String uuid) {
+            return String.join(
+                SEPARATOR,
+                METADATA_PREFIX,
+                Long.toString(primaryTerm),
+                Long.toString(generation, Character.MAX_RADIX),
+                uuid
+            );
+        }
+
+        // Visible for testing
+        static long getPrimaryTerm(String[] filenameTokens) {
+            return Long.parseLong(filenameTokens[1]);
+        }
+
+        // Visible for testing
+        static long getGeneration(String[] filenameTokens) {
+            return Long.parseLong(filenameTokens[2], Character.MAX_RADIX);
+        }
+
+        // Visible for testing
+        static String getUuid(String[] filenameTokens) {
+            return filenameTokens[3];
+        }
+    }
+
+    /**
+     * Returns list of all the segment files uploaded to remote segment store till the last refresh checkpoint.
+     * Any segment file that is uploaded without corresponding metadata file will not be visible as part of listAll().
+     * We chose not to return cache entries for listAll as cache can have entries for stale segments as well.
+     * Even if we plan to delete stale segments from remote segment store, it will be a periodic operation.
+     * @return segment filenames stored in remote segment store
+     * @throws IOException if there were any failures in reading the metadata file
+     */
+    @Override
+    public String[] listAll() throws IOException {
+        return readLatestMetadataFile().keySet().toArray(new String[0]);
+    }
+
+    /**
+     * Delete segment file from remote segment store.
+     * @param name the name of an existing segment file in local filesystem.
+     * @throws IOException if the file exists but could not be deleted.
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        String remoteFilename = getExistingRemoteFilename(name);
+        if (remoteFilename != null) {
+            remoteDataDirectory.deleteFile(remoteFilename);
+            segmentsUploadedToRemoteStore.remove(name);
+        }
+    }
```

Comment:
One rare race could be we delete file and `segmentsUploadedToRemoteStore` hasn't been updated and then a concurrent upload happens and we check for `containsFile` where the stale entry that the file is uploaded is present, will cause the new segment file to get skipped from being uploaded. 
Can you confirm if there is a valid scenario that might hit this race?

Reply:
I don't think of a scenario where this should happen due to following reasons:
1. Segment file with the same name is not created, even after segments are deleted as part of merge, new segment files will always be prefixed with next generation number.
2. Even if segment file with the same name is re-created (this contradicts with #1), we make sure each segment file is uploaded with the unique suffix.